### ISSUE TITLE
v2.2 mem usage keep increasing till OOM when tail large number of files

### ISSUE THREAD

**Issue by yifeng-cerebras** [fluent/fluent-bit#8343]
## Bug Report

**Describe the bug**

on fluentbit v2.2 in k8s, monitoring one NFS folder with all log files in it. 
The logs will not rotate and only be purged after a few days.
As observation, when a new job starts generating 100+ new logs, the memory bump and never reduces **even when it's idle** and after a few jobs, it will OOM.
Am I missing something obvious?

In the screenshot below, it's bumping from 65MB to 110MB after 150 new files generated.
![image](https://github.com/fluent/fluent-bit/assets/102546344/d60ad5b2-328a-466b-9875-c28daddb29d3)

It's output from tail to loki and I can confirm no delay or retry on loki side since the filtered records rate are pretty low.
```
    [PARSER]
        # this parser adds the time key + parse json into kv map
        Name        rt_metric_parser
        Format      json
        Time_Key    timestamp
        Time_Format %s.%L
    [PARSER]
        # this parser parses the file path into job + ns kv map
        name        jobid_ns_parser
        Format      regex
        Regex       .*\/logs\/(?<namespace>[^\/]+)\/(?<wsjob>wsjob[^\/]*)\/(?<replica_type>[^-]*)-(?<replica_id>\d+).*
  fluent-bit.conf: |
    [SERVICE]
        Flush              5
        Parsers_File       parsers.conf
        Parsers_File       custom_parsers.conf
        HTTP_Server        On

    [INPUT]
        Name               tail
        Path               /cb/logs/*/*/*/*.out
        parser             rt_metric_parser
        Tag                rt_metric_tag
        Path_Key           file_path_key
        Refresh_Interval   300
        Mem_Buf_Limit      50MB
        Skip_Long_Lines    On
        Skip_Empty_Lines   On
        Ignore_Older       1h
        Inotify_Watcher    true

    [FILTER]
        # this grep ensures only successfully parsed events (i.e. non-empty "metric" key) go through
        # space required, format: key val
        Name           grep
        Match          rt_metric_tag
        regex          metric .+
    [FILTER]
        # this parser parses the file path into job + ns kv map
        Name           parser
        Match          rt_metric_tag
        Key_Name       file_path_key
        Parser         jobid_ns_parser
        Reserve_Data   On

    [OUTPUT]
        Name                        loki
        Match                       rt_metric_tag
        Host                        100.67.227.255
        Port                        3100
        Retry_Limit                 2
        Net.Keepalive_Idle_Timeout  3600
        Labels                      $metric, $namespace, $wsjob, $replica_type, $replica_id
```

I tried to dump the usage after the bump but I don't see anything. 
```
[2024/01/03 15:29:25] [engine] caught signal (SIGCONT)
[2024/01/03 15:29:25] Fluent Bit Dump

===== Input =====
tail.0 (tail)
│
├─ status
│  └─ overlimit     : no
│     ├─ mem size   : 42.0K (43012 bytes)
│     └─ mem limit  : 47.7M (50000000 bytes)
│
├─ tasks
│  ├─ total tasks   : 3
│  ├─ new           : 0
│  ├─ running       : 3
│  └─ size          : 42.0K (43012 bytes)
│
└─ chunks
   └─ total chunks  : 3
      ├─ up chunks  : 3
      ├─ down chunks: 0
      └─ busy chunks: 3
         ├─ size    : 42.0K (43012 bytes)
         └─ size err: 0


===== Storage Layer =====
total chunks     : 3
├─ mem chunks    : 3
└─ fs chunks     : 0
   ├─ up         : 0
   └─ down       : 0
```

**To Reproduce**
- Rubular link if applicable:
- Example log message if applicable:
```
{"metric":"rt_iter_perf","it":646461.88,"ct":0.001,"timestamp":"1704015101.515451"}
```

- Steps to reproduce the problem:
Batch generating new logs under same folder

**Expected behavior**

Mem usage in reasonable range if inputs bandwidth low  

**Screenshots**

**Your Environment**

* Version used: 2.2.0
* Configuration: default
* Environment name and version (e.g. Kubernetes? What version?): k8s 1.24.4
* Server type and version:
* Operating System and version: Linux sc-r15ra18-s1 4.18.0-348.el8.0.2.x86_64 
* Filters and plugins: tail + grep + parser + loki

**Additional context**

**Comment by yifeng-cerebras** [comment]
Hi, thanks for reply.

> Is this reproducible with the latest image? (fluent/fluent-bit:latest)


I haven't got a chance to test with latest image yet but I did some extra tests to validate the theory that it's related with monitor file list keeps expanding.
From the code, looks like those monitoring list will only be removed after log rotated or deleted which is not happening in my use case(no rotation and files only got deleted after around 7 days).
I tried manually delete those 150 files shortly after tailing done and did see drop in mem usage though seems smaller than the bump. Will get back with more data points.

> Does this happen right after adding 100+ files?

yes, as you can see from screenshot.

> Can you share full Fluent Bit's log?

Will get back soon.

> What's your memory limit?

Tried 1G and 2G, both eventually OOM

> Have you verified the resources actually used by the pod?

I don't have a metrics server installed so unable to do "top" but I [believe the metric for "top"](https://github.com/kubernetes-sigs/metrics-server/blob/d4432d67b2fc435b9c71a89c13659882008a4c54/pkg/sources/summary/summary.go#L206) is the same of "WorkingSetBytes/WSS" as shown on the above screenshot. (65MB to 110MB)


> What's the total number of files opened?

As you can see from the screenshot, it's bumping from around 150 to 300
it's a new test install so total number of files are low.


Just to add:
I've tested multiple times and also on multiple version from 1.9 to 2.2 and reproducing the same issue consistently seeing the mem bump at the new job with 100+ files open.
I'm trying to add rotation to workaround with it but if my theory is right that it's due to monitor list keep growing(though I'm not sure why it's taking so much memory), seems maybe a new config like "stop_monitoring_older" similar as "ignore_older" may help.
I'll come back with more tests and data/log.

**Comment by tmxkaos112** [comment]
@lecaros 
I also have the same problem.

I am currently using version 2.1.10 and inputting using tail.

In my test environment, one log file is created per transaction.

Therefore, numerous log files are created, and I would like to dereference the files that have already been read, but that does not work.

So it seems like the memory continues to increase.

In fact, I entered debug mode, checked that the file was referenced through losf, and deleted the referenced file, and only then was the memory returned.

Is there a solution for this??

Below is my input tail settings.

[INPUT]
   Name       tail
   Path         /fluent-bit/log/*.log
   Tag          logs.*
   Refresh_Interval  1
   Ignore_Older 1m
   Buffer_Chunk_Size 2m
   Buffer_Max_Size   100m
   Skip_Long_Lines   Off
   Skip_Empty_Lines On
   Mem_Buf_Limit 512m

**Comment by yifeng-cerebras** [comment]
I workaround by adding rotation policy to ensure the monitor list get cleaned after some interval.
and validated that mem usage is reduced after logs rotation but will kept if never rotated
also, it seems the mem bump issue is bigger using NFS.

It seems not relative common use case so think there can be an option to allow auto remove those files if no change after some interval like `stop_monitoring_idle_after`

![image](https://github.com/fluent/fluent-bit/assets/102546344/9f3835b2-0132-480e-9578-0329085a0245)

**Comment by tmxkaos112** [comment]
@yifeng-cerebras 

Did you create a script that deletes the log file at certain intervals?
Can you share how you solved it?

**Comment by yifeng-cerebras** [comment]
@tmxkaos112 
in our use case, we are on k8s, so we redirect logs to pod stdout and k8s has auto rotation enabled
but think if you can also do it manually or through tools like [logrotate](https://www.baeldung.com/linux/rotating-logs-logrotate)

from the code, looks like monitor list will only be updated after [file rotated/deleted](https://github.com/fluent/fluent-bit/blob/master/plugins/in_tail/tail_file.c#L1803) 
so basically if you move/delete the file periodically, the list can be cleaned

**Comment by nr-rkallempudi** [comment]
> [@lecaros](https://github.com/lecaros) I also have the same problem.
> 
> I am currently using version 2.1.10 and inputting using tail.
> 
> In my test environment, one log file is created per transaction.
> 
> Therefore, numerous log files are created, and I would like to dereference the files that have already been read, but that does not work.
> 
> So it seems like the memory continues to increase.
> 
> In fact, I entered debug mode, checked that the file was referenced through lsof, and deleted the referenced file, and only then was the memory returned.
> 
> Is there a solution for this??
> 
> Below is my input tail settings.
> 
> [INPUT] Name tail Path /fluent-bit/log/_.log Tag logs._ Refresh_Interval 1 Ignore_Older 1m Buffer_Chunk_Size 2m Buffer_Max_Size 100m Skip_Long_Lines Off Skip_Empty_Lines On Mem_Buf_Limit 512m


In your specific case, where a different log file created for every transaction or something, can't you tune the INPUT tail plugin with configurable parameter **Exit_On_Eof** = True?

`Exit_On_Eof : When reading a file will exit as soon as it reach the end of the file. Useful for bulk load and tests`

https://docs.fluentbit.io/manual/pipeline/inputs/tail
